### PR TITLE
Feature/postpone the read write split decision

### PR DESCRIPTION
**Pull request by newborn22**
## Description
1. code about resolve SQL routing tablet type is reconstructed, now they are moved to the func `ResolveTabletType`. To achieve this, a new struct field `ResolverOptions` is added to safesession.

2. Redundant calls to Parse2 is removed by calling Parse2 outside the func newVCursorImpl and the func getPlan.

3. If a SQL is not in read only transaction, and suggested tablet type of the SQL is REPLICA, and neither keyspace tablet type nor user tablet type hint is set, then load balance module will pick the final tablet between REPLICA and RDONLY tablets.
 
4. If Keyspace tablet type or user tablet type hint is set as RDONLY, now the vtgate will route SQLs to RDONLY tablets instead of REPLICA tablets.

5. The command about SQL execution vttablet info display is change from `set enable_display_sql_execution_vttablet_type=true` to `set enable_display_sql_execution_vttablet=true`, and the content is enriched, now it looks like:
```
mysql> select * from t;
+------+------+
| c1   | c2   |
+------+------+
|    2 |    2 |
| 2222 | 2222 |
+------+------+
2 rows in set (0.01 sec)
the sql is executed on zone1-101(PRIMARY)
suggested vttablet type is PRIMARY, read write splitting ratio is 100%, read write splitting policy is DISABLE, sql should execute on PRIMARY
load balance policy is RANDOM, load balance between PRIMARY vttablets

mysql> set read_write_splitting_policy='random';
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t;
+------+------+
| c1   | c2   |
+------+------+
|    2 |    2 |
| 2222 | 2222 |
+------+------+
2 rows in set (0.02 sec)
the sql is executed on zone1-100(REPLICA)
suggested vttablet type is REPLICA, read write splitting ratio is 100%, read write splitting policy is random, and is a pure select sql
load balance policy is RANDOM, load balance between REPLICA and RDONLY vttablets

```
## Related Issue(s)
#117 
#369 

## Checklist

-  e2e tests TestRoutingTabletType in file routing_test.go were added in wesql cluster tests

### CODE REVIEW

File: go/vt/vtgate/read_write_splitter_test.go [hunk 1]
```diff
@@ -380,32 +380,32 @@ func Test_suggestTabletType_random(t *testing.T) {
...
+		//		readWriteSplittingRatio:  int32(70),
+		//		inTransaction:            false,
+		//		hasCreatedTempTables:     false,
+		//		hasAdvisoryLock:          false,
+		//		sql:                      "SELECT * FROM users;",
+		//	},
+		//	wantRatio: 0,
+		//	wantErr:   assert.NoError,
+		//},
 		{
```
Comment (earayu) [inline go/vt/vtgate/read_write_splitter_test.go:409]: why comment out?

File: go/vt/vtgate/scatter_conn.go [hunk 2]
```diff
@@ -227,6 +223,14 @@ func (stc *ScatterConn) ExecuteMultiShard(
...
 				opts.LoadBalancePolicy = schema.ToLoadBalancePolicy(session.GetReadWriteSplittingPolicy())
 				opts.AccountVerificationEnabled = mysqlAuthServerImpl != global.AuthServerNone
+				// if the target tablet type is REPLIC or RDONLY and it's not set by user, then the load balance should decide between REPLIC and RDONLY nodes
+				if (rs.Target.TabletType == topodatapb.TabletType_REPLICA || rs.Target.TabletType == topodatapb.TabletType_RDONLY) &&
+					session.ResolverOptions.UserHintTabletType == topodatapb.TabletType_UNKNOWN &&
+					session.ResolverOptions.KeyspaceTabletType == topodatapb.TabletType_UNKNOWN &&
+					!session.ResolverOptions.ShouldForceRouteToReadOnlyTxnVttablet {
+					// if ShouldForceRouteToReadOnly is true, it means it is in read only txn, and sql should be routed to one single node
+					opts.CanLoadBalanceBetweenReplicAndRdonly = true
+				}
```
Comment (earayu) [inline go/vt/vtgate/scatter_conn.go:233]: Curious about the logic here

File: go/vt/vtgate/plan_execute.go [hunk 3]
```diff
@@ -315,3 +330,80 @@ func interceptDMLWithoutWhereEnable(safeSession *SafeSession, stmt sqlparser.Sta
...
+	if err != nil {
+		return err
+	}
+	// get UserHintTabletType
+	safeSession.ResolverOptions.UserHintTabletType, err = GetUserHintTabletType(stmt, vcursor)
+	if err != nil {
+		return err
+	}
+	// decide tablet type
+	vcursor.tabletType = ResolveTabletTypeBaseOnOptions(vcursor.safeSession.ResolverOptions)
```
[the remark this answers is not in the document]
Comment (newborn22) [reply go/vt/vtgate/plan_execute.go]: Yes,  we can directly return if tablet type with higher priority is set.

File: go/vt/vtgate/plan_execute.go [hunk 4]
```diff
@@ -315,3 +330,80 @@ func interceptDMLWithoutWhereEnable(safeSession *SafeSession, stmt sqlparser.Sta
...
+			SuggestedTabletType:                   topodatapb.TabletType_UNKNOWN,
+			ShouldForceRouteToReadOnlyTxnVttablet: false}
+	}
+}
+
+func GetSuggestedTabletType(safeSession *SafeSession, sql string) (topodatapb.TabletType, error) {
+	// current sql is in read only transaction or not, if is a sql begins a read-only tx like "start transaction read only;", the isReadOnlyTx will be false,
+	// but the sql will not send to tablet to execute
+	isReadOnlyTx := safeSession.Session.InTransaction && safeSession.Session.TransactionAccessMode == vtgatepb.TransactionAccessMode_READ_ONLY
+	// user can use set @EnableReadWriteSplitForReadOnlyTxn=true/false to enable/disable read only tx routing
```
[the remark this answers is not in the document]
Comment (newborn22) [reply go/vt/vtgate/plan_execute.go]: Yes

File: go/vt/vtgate/plan_execute.go [hunk 5]
```diff
@@ -315,3 +330,80 @@ func interceptDMLWithoutWhereEnable(safeSession *SafeSession, stmt sqlparser.Sta
...
+			KeyspaceTabletType:                    topodatapb.TabletType_UNKNOWN,
+			UserHintTabletType:                    topodatapb.TabletType_UNKNOWN,
+			SuggestedTabletType:                   topodatapb.TabletType_UNKNOWN,
+			ShouldForceRouteToReadOnlyTxnVttablet: false}
+	}
+}
+
+func GetSuggestedTabletType(safeSession *SafeSession, sql string) (topodatapb.TabletType, error) {
+	// current sql is in read only transaction or not, if is a sql begins a read-only tx like "start transaction read only;", the isReadOnlyTx will be false,
+	// but the sql will not send to tablet to execute
```
[the remark this answers is not in the document]
Comment (newborn22) [reply go/vt/vtgate/plan_execute.go]: Modified

**Review by earayu — APPROVED** [review]
LGTM

**Pull request merged**